File: csharp/demo/demo19-read-write-lock.cs
/*
 * READ-WRITE LOCKS
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;



class Demo19 : IRunnable
{
    public void run()
    {
        var rwlk = new ReaderWriterLock();


        const int NUM_THREADS_READ = 10;
        const int NUM_THREADS_WRITE = 4;
        const int NUM_ARGS = 3;


        var arg = Enumerable.Range(0, NUM_ARGS).ToArray();
        var rand = new Random();

        var lstThRead = new List<Thread>();
        var lstThWrite = new List<Thread>();


        for (int i = 0; i < NUM_THREADS_READ; ++i)
        {
            lstThRead.Add(new Thread(() =>
            {
                int timeWait = arg[ rand.Next(arg.Length) ];
                Thread.Sleep(1000 * timeWait);

                // Should catch exception
                rwlk.AcquireReaderLock(1000);

                Console.WriteLine("read: " + Resource.value);

                rwlk.ReleaseReaderLock();
            }));
        }


        for (int i = 0; i < NUM_THREADS_WRITE; ++i)
        {
            lstThWrite.Add(new Thread(() =>
            {
                int timeWait = arg[ rand.Next(arg.Length) ];
                Thread.Sleep(1000 * timeWait);

                // Should catch exception
                rwlk.AcquireWriterLock(1000);

                Resource.value = rand.Next(100);
                Console.WriteLine("write: " + Resource.value);

                rwlk.ReleaseWriterLock();
            }));
        }


        lstThRead.ForEach(th => th.Start());
        lstThWrite.ForEach(th => th.Start());
    }



    class Resource
    {
        public static volatile int value;
    }
}

File: python/demo19_read_write_lock.py
'''
READ-WRITE LOCKS
Read-write locks in Python are not supported by default.
So, I use mylib_rwlock for this demonstration.
'''

import time
import random
import threading
from mylib_rwlock import ReadWriteLock



rwlock = ReadWriteLock()
resource = 0



def read_func(wait_time):
    time.sleep(wait_time)

    with rwlock.readlock():
        print(f'read: {resource}')



def write_func(wait_time):
    global resource
    time.sleep(wait_time)

    with rwlock.writelock():
        resource = random.randint(0, 99)
        print(f'write: {resource}')



NUM_THREADS_READ = 10
NUM_THREADS_WRITE = 4
TIME_WAIT_MAX = 2

lstth_read = [
    threading.Thread(target=read_func, args=(random.randint(0, TIME_WAIT_MAX),))
    for _ in range(NUM_THREADS_READ)
]

lstth_write = [
    threading.Thread(target=write_func, args=(random.randint(0, TIME_WAIT_MAX),))
    for _ in range(NUM_THREADS_WRITE)
]

for th in lstth_read:
    th.start()

for th in lstth_write:
    th.start()
